revive: rules's severity config has no effect

### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

I tried to configure specific severity for some revive's rules but they are completely ignored.
I tried with revive directly and it works perfectly with this config file:
```toml
#  When set to false, ignores files with "GENERATED" header, similar to golint
ignoreGeneratedHeader = true

# Sets the default severity to "warning"
severity = "error"

# Sets the default failure confidence. This means that linting errors
# with less than 0.8 confidence will be ignored.
confidence = 0.8

# Sets the error code for failures with severity "error"
errorCode = 0

# Sets the error code for failures with severity "warning"
warningCode = 0

# Configuration of the `cyclomatic` rule. Here we specify that
# the rule should fail if it detects code with higher complexity than 10.
[rule.var-naming]
  severity = "warning"
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.48.0 built from 2d8fea81 on 2022-08-04T18:44:38Z
```

</details>

### Configuration file

<details>

```yaml
# Golangci-lint documentation: https://golangci-lint.run/

# Analysis running options
run:
  # Timeout for analysis
  timeout: 5m

  # Disable tests to avoid redundancy with test workflow
  tests: false
  modules-download-mode: vendor

issues:
  # Maximum issues count per one linter
  max-issues-per-linter: 0
  # Maximum count of issues with the same text
  max-same-issues: 0

severity:
  default-severity: error
  case-sensitive: true
  rules:
    - linters:
        - godox
        - goconst
        - lll
        - misspell
      severity: warning

# Configuration of specific linters
linters-settings:
  goconst:
    min-len: 2
    min-occurrences: 3
  gosimple:
    go: "1.18"
    checks: ["all"]
  govet:
    check-shadowing: true
    settings:
      printf:
        funcs:
          - (github.com/sirupsen/logrus).Debugf
          - (github.com/sirupsen/logrus).Infof
          - (github.com/sirupsen/logrus).Warnf
          - (github.com/sirupsen/logrus).Errorf
          - (github.com/sirupsen/logrus).Fatalf
  misspell:
    locale: US
  lll:
    line-length: 140
  revive:
    ignore-generated-header: false
    severity: error
    confidence: 0.8
    rules:
      - name: blank-imports
        severity: warning
      - name: context-as-argument
        severity: warning
      - name: context-keys-type
        severity: warning
      - name: cyclomatic
        severity: warning
        arguments:
          - 30 # Maximum cyclomatic complexity
      - name: error-return
        severity: warning
      - name: error-strings
        severity: warning
      - name: error-naming
        severity: warning
      - name: exported
        severity: warning
      - name: if-return
        severity: warning
      - name: increment-decrement
        severity: warning
      - name: var-naming
        severity: warning
      - name: var-declaration
        severity: warning
      - name: package-comments
        severity: warning
      - name: range
        severity: warning
      - name: receiver-naming
        severity: warning
      - name: time-naming
        severity: warning
      - name: unexported-return
        severity: warning
      - name: indent-error-flow
        severity: warning
      - name: errorf
        severity: warning
      - name: empty-block
        severity: warning
      - name: superfluous-else
        severity: warning
      - name: unreachable-code
        severity: warning
      - name: redefines-builtin-id
        severity: warning
  staticcheck:
    go: "1.18"
    checks: ["all"]

linters:
  disable-all: true
  enable:
    - deadcode
    - errcheck
    - exportloopref
    - goconst
    - godox
    - gofmt
    - goimports
    - gosec
    - gosimple
    - govet
    - ineffassign
    - lll
    - misspell
    - prealloc
    - revive
    - rowserrcheck
    - staticcheck
    - typecheck
    - unconvert
    - unparam
    - unused
    - varcheck
    - whitespace

```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.18.3 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/hbollon/.cache/go-build"
GOENV="/home/hbollon/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/hbollon/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/hbollon/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18.3"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="0"
GOMOD="/home/hbollon/git/camptocamp/github.com/pulumi-aws-schweizmobil/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3768830388=/tmp/go-build -gno-record-gcc-switches"

```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/hbollon/git/camptocamp/github.com/pulumi-aws-schweizmobil /home/hbollon/git/camptocamp/github.com /home/hbollon/git/camptocamp /home/hbollon/git /home/hbollon /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 23 linters: [deadcode errcheck exportloopref goconst godox gofmt goimports gosec gosimple govet ineffassign lll misspell prealloc revive rowserrcheck staticcheck typecheck unconvert unparam unused varcheck whitespace] 
INFO [loader] Go packages loading at mode 575 (exports_file|imports|name|types_sizes|compiled_files|deps|files) took 297.133506ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 3.516474ms 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
WARN [linter] revive: the following rules (context-keys-type,errorf,time-naming,unexported-return,var-declaration) are ignored due to a performance problem (https://github.com/golangci/golangci-lint/issues/2997) 
INFO [linters context/goanalysis] analyzers took 28.882281765s with top 10 stages: buildir: 22.87677273s, inspect: 1.227991458s, fact_deprecated: 961.468251ms, ctrlflow: 795.982667ms, printf: 661.610435ms, buildssa: 460.026843ms, nilness: 415.28584ms, fact_purity: 410.979807ms, SA5012: 278.633748ms, typedness: 272.857575ms 
WARN [linters context] rowserrcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
INFO [runner] Issues before processing: 54, after processing: 8 
INFO [runner] Processors filtering stat (out/in): autogenerated_exclude: 54/54, severity-rules-case-sensitive: 8/8, path_prefixer: 8/8, sort_results: 8/8, cgo: 54/54, skip_files: 54/54, uniq_by_line: 8/8, source_code: 8/8, path_shortener: 8/8, filename_unadjuster: 54/54, identifier_marker: 54/54, exclude: 54/54, diff: 8/8, max_same_issues: 8/8, path_prettifier: 54/54, exclude-rules: 8/54, nolint: 8/8, max_per_file_from_linter: 8/8, max_from_linter: 8/8, skip_dirs: 54/54 
INFO [runner] processing took 5.025759ms with stages: exclude-rules: 2.342011ms, identifier_marker: 1.184585ms, nolint: 1.026599ms, path_prettifier: 226.442µs, autogenerated_exclude: 115.616µs, skip_dirs: 57.544µs, source_code: 55.683µs, uniq_by_line: 4.459µs, cgo: 3.697µs, path_shortener: 3.105µs, filename_unadjuster: 2.186µs, severity-rules-case-sensitive: 1.212µs, max_per_file_from_linter: 987ns, max_same_issues: 546ns, skip_files: 229ns, exclude: 221ns, diff: 187ns, max_from_linter: 182ns, sort_results: 161ns, path_prefixer: 107ns 
INFO [runner] linters took 6.877429384s with stages: goanalysis_metalinter: 6.872331656s, rowserrcheck: 6.736µs 
internal/pkg/components/foundations/dns.go:42:28: Error return value of `f.errorWrapper.Wrap` is not checked (errcheck)
		defer f.errorWrapper.Wrap(err)
		                         ^
internal/pkg/components/foundations/ecr.go:67: internal/pkg/components/foundations/ecr.go:67: Line contains TODO/BUG/FIXME: "TODO" (godox)
			// TODO
internal/pkg/components/foundations/ecr.go:72: internal/pkg/components/foundations/ecr.go:72: Line contains TODO/BUG/FIXME: "TODO" (godox)
				// TODO
internal/pkg/components/foundations/ecr.go:113: internal/pkg/components/foundations/ecr.go:113: Line contains TODO/BUG/FIXME: "TODO" (godox)
			// TODO
internal/pkg/components/foundations/ecr.go:118: internal/pkg/components/foundations/ecr.go:118: Line contains TODO/BUG/FIXME: "TODO" (godox)
				// TODO
internal/pkg/components/foundations/network.go:107: internal/pkg/components/foundations/network.go:107: Line contains TODO/BUG/FIXME: "TODO" (godox)
				// TODO
internal/pkg/components/foundations/dns.go:13:23: var-naming: method newDnsZone should be newDNSZone (revive)
func (f *Foundations) newDnsZone(ctx *pulumi.Context) (err error) {
                      ^
internal/pkg/components/foundations/foundations.go:85:6: exported: type name will be used as foundations.FoundationsArgs by other packages, and that stutters; consider calling this Args (revive)
type FoundationsArgs struct {
     ^
INFO File cache stats: 19 entries of total size 61.7KiB 
INFO Memory: 73 samples, avg is 462.4MB, max is 667.2MB 
INFO Execution took 7.182767414s
```

</details>

Additionally, there is severity informations parsed with jq:
<details>

```console
$ golangci-lint run --out-format json | jq  '.Issues[] | {text: .Text, severity: .Severity }'
WARN [linters context] rowserrcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
{
  "text": "Error return value of `f.errorWrapper.Wrap` is not checked",
  "severity": "error"
}
{
  "text": "internal/pkg/components/foundations/ecr.go:67: Line contains TODO/BUG/FIXME: \"TODO\"",
  "severity": "warning"
}
{
  "text": "internal/pkg/components/foundations/ecr.go:72: Line contains TODO/BUG/FIXME: \"TODO\"",
  "severity": "warning"
}
{
  "text": "internal/pkg/components/foundations/ecr.go:113: Line contains TODO/BUG/FIXME: \"TODO\"",
  "severity": "warning"
}
{
  "text": "internal/pkg/components/foundations/ecr.go:118: Line contains TODO/BUG/FIXME: \"TODO\"",
  "severity": "warning"
}
{
  "text": "internal/pkg/components/foundations/network.go:107: Line contains TODO/BUG/FIXME: \"TODO\"",
  "severity": "warning"
}
{
  "text": "var-naming: method newDnsZone should be newDNSZone",
  "severity": "error"
}
{
  "text": "exported: type name will be used as foundations.FoundationsArgs by other packages, and that stutters; consider calling this Args",
  "severity": "error"
}
```

</details>

`var-naming` and `exported` should be at **warning** level according to my configuration file...

### Code example or link to a public repository

<details>

```go
package main

func newDnsZone() error {
	return nil
}

func main() {
	newDnsZone()
}

```

</details>

With my config it should return:
```console
$ golangci-lint run --out-format json | jq  '.Issues[] | {text: .Text, severity: .Severity }'
WARN [linters context] rowserrcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
{
  "text": "Error return value is not checked",
  "severity": "error"
}
{
  "text": "var-naming: func newDnsZone should be newDNSZone",
  "severity": "error"
}
```

Which is not the intended result since the `var-naming` rule appear as error and not warning

--- a/.golangci.reference.yml
+++ b/.golangci.reference.yml
@@ -2881,6 +2881,8 @@ severity:
   # - GitHub: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message
   # - TeamCity: https://www.jetbrains.com/help/teamcity/service-messages.html#Inspection+Instance
   #
+  # `@` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
+  #
   # Default: ""
   default-severity: error
 
@@ -2891,6 +2893,9 @@ severity:
   # When a list of severity rules are provided, severity information will be added to lint issues.
   # Severity rules have the same filtering capability as exclude rules
   # except you are allowed to specify one matcher per severity rule.
+  #
+  # `@` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
+  #
   # Only affects out formats that support setting severity information.
   #
   # Default: []

--- a/.golangci.reference.yml
+++ b/.golangci.reference.yml
@@ -2888,10 +2888,6 @@ severity:
   # Default: false
   case-sensitive: true
 
-  # Don't override severity defined by linters.
-  # Default: false
-  keep-linter-severity: true
-
   # When a list of severity rules are provided, severity information will be added to lint issues.
   # Severity rules have the same filtering capability as exclude rules
   # except you are allowed to specify one matcher per severity rule.

--- a/.golangci.reference.yml
+++ b/.golangci.reference.yml
@@ -2881,7 +2881,7 @@ severity:
   # - GitHub: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message
   # - TeamCity: https://www.jetbrains.com/help/teamcity/service-messages.html#Inspection+Instance
   #
-  # `@` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
+  # `@linter` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
   #
   # Default: ""
   default-severity: error
@@ -2894,7 +2894,7 @@ severity:
   # Severity rules have the same filtering capability as exclude rules
   # except you are allowed to specify one matcher per severity rule.
   #
-  # `@` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
+  # `@linter` can be used as severity value to keep the severity from linters (e.g. revive, gosec, ...)
   #
   # Only affects out formats that support setting severity information.
   #

--- a/pkg/config/severity.go
+++ b/pkg/config/severity.go
@@ -8,10 +8,9 @@ import (
 const severityRuleMinConditionsCount = 1
 
 type Severity struct {
-	Default            string         `mapstructure:"default-severity"`
-	CaseSensitive      bool           `mapstructure:"case-sensitive"`
-	Rules              []SeverityRule `mapstructure:"rules"`
-	KeepLinterSeverity bool           `mapstructure:"keep-linter-severity"` // TODO(ldez): in v2 should be changed to `Override`.
+	Default       string         `mapstructure:"default-severity"`
+	CaseSensitive bool           `mapstructure:"case-sensitive"`
+	Rules         []SeverityRule `mapstructure:"rules"`
 }
 
 func (s *Severity) Validate() error {

--- a/pkg/lint/runner.go
+++ b/pkg/lint/runner.go
@@ -302,7 +302,6 @@ func getSeverityRulesProcessor(cfg *config.Severity, log logutils.Log, files *fs
 		Default:       cfg.Default,
 		Rules:         severityRules,
 		CaseSensitive: cfg.CaseSensitive,
-		Override:      !cfg.KeepLinterSeverity,
 	}
 
 	return processors.NewSeverity(log.Child(logutils.DebugKeySeverityRules), files, severityOpts)

--- a/pkg/result/processors/processor_test.go
+++ b/pkg/result/processors/processor_test.go
@@ -22,6 +22,7 @@ func newIssueFromIssueTestCase(c issueTestCase) result.Issue {
 	return result.Issue{
 		Text:       c.Text,
 		FromLinter: c.Linter,
+		Severity:   c.Severity,
 		Pos: token.Position{
 			Filename: c.Path,
 			Line:     c.Line,

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -66,6 +66,10 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 			rule := rule
 
 			if rule.match(issue, p.files, p.log) {
+				if rule.severity == "@" || rule.severity == "" && p.defaultSeverity == "@" {
+					return issue
+				}
+
 				issue.Severity = rule.severity
 				if issue.Severity == "" {
 					issue.Severity = p.defaultSeverity
@@ -75,7 +79,9 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 			}
 		}
 
-		issue.Severity = p.defaultSeverity
+		if p.defaultSeverity != "@" {
+			issue.Severity = p.defaultSeverity
+		}
 
 		return issue
 	}), nil

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -66,9 +66,9 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 			rule := rule
 
 			if rule.match(issue, p.files, p.log) {
-				issue.Severity = p.defaultSeverity
-				if rule.severity != "" {
-					issue.Severity = rule.severity
+				issue.Severity = rule.severity
+				if issue.Severity == "" {
+					issue.Severity = p.defaultSeverity
 				}
 
 				return issue

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -68,8 +68,6 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 
 func (p *Severity) transform(issue *result.Issue) *result.Issue {
 	for _, rule := range p.rules {
-		rule := rule
-
 		if rule.match(issue, p.files, p.log) {
 			if rule.severity == severityFromLinter || rule.severity == "" && p.defaultSeverity == severityFromLinter {
 				return issue

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -66,12 +66,11 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 			rule := rule
 
 			if rule.match(issue, p.files, p.log) {
-				ruleSeverity := p.defaultSeverity
+				issue.Severity = p.defaultSeverity
 				if rule.severity != "" {
-					ruleSeverity = rule.severity
+					issue.Severity = rule.severity
 				}
 
-				issue.Severity = ruleSeverity
 				return issue
 			}
 		}

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -65,12 +65,12 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 		for _, rule := range p.rules {
 			rule := rule
 
-			ruleSeverity := p.defaultSeverity
-			if rule.severity != "" {
-				ruleSeverity = rule.severity
-			}
-
 			if rule.match(issue, p.files, p.log) {
+				ruleSeverity := p.defaultSeverity
+				if rule.severity != "" {
+					ruleSeverity = rule.severity
+				}
+
 				issue.Severity = ruleSeverity
 				return issue
 			}

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -61,30 +61,32 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 		return issues, nil
 	}
 
-	return transformIssues(issues, func(issue *result.Issue) *result.Issue {
-		for _, rule := range p.rules {
-			rule := rule
+	return transformIssues(issues, p.transform), nil
+}
 
-			if rule.match(issue, p.files, p.log) {
-				if rule.severity == "@" || rule.severity == "" && p.defaultSeverity == "@" {
-					return issue
-				}
+func (p *Severity) transform(issue *result.Issue) *result.Issue {
+	for _, rule := range p.rules {
+		rule := rule
 
-				issue.Severity = rule.severity
-				if issue.Severity == "" {
-					issue.Severity = p.defaultSeverity
-				}
-
+		if rule.match(issue, p.files, p.log) {
+			if rule.severity == "@" || rule.severity == "" && p.defaultSeverity == "@" {
 				return issue
 			}
-		}
 
-		if p.defaultSeverity != "@" {
-			issue.Severity = p.defaultSeverity
-		}
+			issue.Severity = rule.severity
+			if issue.Severity == "" {
+				issue.Severity = p.defaultSeverity
+			}
 
-		return issue
-	}), nil
+			return issue
+		}
+	}
+
+	if p.defaultSeverity != "@" {
+		issue.Severity = p.defaultSeverity
+	}
+
+	return issue
 }
 
 func (p *Severity) Name() string { return p.name }

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -8,6 +8,8 @@ import (
 	"github.com/golangci/golangci-lint/pkg/result"
 )
 
+const severityFromLinter = "@linter"
+
 var _ Processor = &Severity{}
 
 type severityRule struct {
@@ -69,7 +71,7 @@ func (p *Severity) transform(issue *result.Issue) *result.Issue {
 		rule := rule
 
 		if rule.match(issue, p.files, p.log) {
-			if rule.severity == "@" || rule.severity == "" && p.defaultSeverity == "@" {
+			if rule.severity == severityFromLinter || rule.severity == "" && p.defaultSeverity == severityFromLinter {
 				return issue
 			}
 
@@ -82,7 +84,7 @@ func (p *Severity) transform(issue *result.Issue) *result.Issue {
 		}
 	}
 
-	if p.defaultSeverity != "@" {
+	if p.defaultSeverity != severityFromLinter {
 		issue.Severity = p.defaultSeverity
 	}
 

--- a/pkg/result/processors/severity.go
+++ b/pkg/result/processors/severity.go
@@ -24,7 +24,6 @@ type SeverityOptions struct {
 	Default       string
 	Rules         []SeverityRule
 	CaseSensitive bool
-	Override      bool
 }
 
 type Severity struct {
@@ -36,7 +35,6 @@ type Severity struct {
 
 	defaultSeverity string
 	rules           []severityRule
-	override        bool
 }
 
 func NewSeverity(log logutils.Log, files *fsutils.Files, opts SeverityOptions) *Severity {
@@ -45,7 +43,6 @@ func NewSeverity(log logutils.Log, files *fsutils.Files, opts SeverityOptions) *
 		files:           files,
 		log:             log,
 		defaultSeverity: opts.Default,
-		override:        opts.Override,
 	}
 
 	prefix := caseInsensitivePrefix
@@ -65,10 +62,6 @@ func (p *Severity) Process(issues []result.Issue) ([]result.Issue, error) {
 	}
 
 	return transformIssues(issues, func(issue *result.Issue) *result.Issue {
-		if issue.Severity != "" && !p.override {
-			return issue
-		}
-
 		for _, rule := range p.rules {
 			rule := rule
 

--- a/pkg/result/processors/severity_test.go
+++ b/pkg/result/processors/severity_test.go
@@ -310,3 +310,187 @@ func TestSeverity_caseSensitive(t *testing.T) {
 
 	assert.Equal(t, expectedCases, resultingCases)
 }
+
+func TestSeverity_transform(t *testing.T) {
+	lineCache := fsutils.NewLineCache(fsutils.NewFileCache())
+	files := fsutils.NewFiles(lineCache, "")
+
+	testCases := []struct {
+		desc     string
+		opts     SeverityOptions
+		issue    *result.Issue
+		expected *result.Issue
+	}{
+		{
+			desc: "apply severity from rule",
+			opts: SeverityOptions{
+				Default: "error",
+				Rules: []SeverityRule{
+					{
+						Severity: "info",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+				Severity:   "info",
+			},
+		},
+		{
+			desc: "apply severity from default",
+			opts: SeverityOptions{
+				Default: "error",
+				Rules: []SeverityRule{
+					{
+						Severity: "info",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "error",
+			},
+		},
+		{
+			desc: "severity from rule override severity from linter",
+			opts: SeverityOptions{
+				Default: "error",
+				Rules: []SeverityRule{
+					{
+						Severity: "info",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+				Severity:   "huge",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+				Severity:   "info",
+			},
+		},
+		{
+			desc: "severity from default override severity from linter",
+			opts: SeverityOptions{
+				Default: "error",
+				Rules: []SeverityRule{
+					{
+						Severity: "info",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "huge",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "error",
+			},
+		},
+		{
+			desc: "keep severity from linter as rule",
+			opts: SeverityOptions{
+				Default: "error",
+				Rules: []SeverityRule{
+					{
+						Severity: "@",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+				Severity:   "huge",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter1",
+				Severity:   "huge",
+			},
+		},
+		{
+			desc: "keep severity from linter as default",
+			opts: SeverityOptions{
+				Default: "@",
+				Rules: []SeverityRule{
+					{
+						Severity: "info",
+						BaseRule: BaseRule{
+							Linters: []string{"linter1"},
+						},
+					},
+				},
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "huge",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "huge",
+			},
+		},
+		{
+			desc: "keep severity from linter as default (without rule)",
+			opts: SeverityOptions{
+				Default: "@",
+			},
+			issue: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "huge",
+			},
+			expected: &result.Issue{
+				Text:       "This is a report",
+				FromLinter: "linter2",
+				Severity:   "huge",
+			},
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			p := NewSeverity(nil, files, test.opts)
+
+			newIssue := p.transform(test.issue)
+
+			assert.Equal(t, test.expected, newIssue)
+		})
+	}
+}

--- a/pkg/result/processors/severity_test.go
+++ b/pkg/result/processors/severity_test.go
@@ -421,7 +421,7 @@ func TestSeverity_transform(t *testing.T) {
 				Default: "error",
 				Rules: []SeverityRule{
 					{
-						Severity: "@",
+						Severity: severityFromLinter,
 						BaseRule: BaseRule{
 							Linters: []string{"linter1"},
 						},
@@ -442,7 +442,7 @@ func TestSeverity_transform(t *testing.T) {
 		{
 			desc: "keep severity from linter as default",
 			opts: SeverityOptions{
-				Default: "@",
+				Default: severityFromLinter,
 				Rules: []SeverityRule{
 					{
 						Severity: "info",
@@ -466,7 +466,7 @@ func TestSeverity_transform(t *testing.T) {
 		{
 			desc: "keep severity from linter as default (without rule)",
 			opts: SeverityOptions{
-				Default: "@",
+				Default: severityFromLinter,
 			},
 			issue: &result.Issue{
 				Text:       "This is a report",

